include/cuco/static_multimap.cuh

```diff
@@ -0,0 +1,1808 @@
+#include <cuco/traits.hpp>
+
+#if defined(CUDART_VERSION) && (CUDART_VERSION >= 11000) && defined(__CUDA_ARCH__) && \
+  (__CUDA_ARCH__ >= 700)
+#define CUCO_HAS_CUDA_BARRIER
+#endif
+
+// cg::memcpy_aysnc is supported for CUDA 11.1 and up
+#if defined(CUDART_VERSION) && (CUDART_VERSION >= 11100)
+#define CUCO_HAS_CG_MEMCPY_ASYNC
+#endif
+
+#if defined(CUCO_HAS_CUDA_BARRIER)
+#include <cuda/barrier>
+#endif
+
+#include <cuco/detail/error.hpp>
+#include <cuco/detail/prime.hpp>
+#include <cuco/detail/probe_sequences.cuh>
+#include <cuco/detail/static_multimap_kernels.cuh>
+
+namespace cuco {
+
+/**
+ * @brief A GPU-accelerated, unordered, associative container of key-value
+ * pairs that supports equivalent keys.
+ *
+ * Allows constant time concurrent inserts or concurrent find operations from threads in device
+ * code. Concurrent insert/find is allowed only when
+ * `static_multimap<Key, Value>::supports_concurrent_insert_find()` is true.
+ *
+ * Current limitations:
+ * - Requires keys and values where `cuco::is_bitwise_comparable<T>::value` is true
+ * - Comparisons against the "sentinel" values will always be done with bitwise comparisons
+ * Therefore, the objects must have unique, bitwise object representations (e.g., no padding bits).
+ * - Does not support erasing keys
+ * - Capacity is fixed and will not grow automatically
+ * - Requires the user to specify sentinel values for both key and mapped value
+ * to indicate empty slots
+ * - Concurrent insert/find is only supported when `static_multimap<Key,
+ * Value>::supports_concurrent_insert_find()` is true`
+ *
+ * The `static_multimap` supports two types of operations:
+ * - Host-side "bulk" operations
+ * - Device-side "singular" operations
+ *
+ * The host-side bulk operations include `insert`, `contains`, `count`, `retrieve` and their
+ * variants. These APIs should be used when there are a large number of keys to insert or lookup in
+ * the map. For example, given a range of keys specified by device-accessible iterators, the bulk
+ * `insert` function will insert all keys into the map.
+ *
+ * The singular device-side operations allow individual threads to perform
+ * independent operations (e.g. `insert`, etc.) from device code. These
+ * operations are accessed through non-owning, trivially copyable "view" types:
+ * `device_view` and `device_mutable_view`. The `device_view` class is an
+ * immutable view that allows only non-modifying operations such as `count` or
+ * `contains`. The `device_mutable_view` class only allows `insert` operations.
+ * The two types are separate to prevent erroneous concurrent insert/find
+ * operations.
+ *
+ * By default, query operations (e.g. `count` and `retrieve`) take `Key` as input and do not include
+ * non-matches in the output. APIs with `_outer` suffix should be used if non-match handling is
+ * desired. The `pair_` prefix indicates that the input data are key-value pairs whose type can be
+ * converted to multimap's `value_type`.
```

Comment:
```suggestion
 * By default, when querying for a Key `k` in operations like `count` or `retrieve`, if `k` is not present in the map, it will  
 * not contribute to the output. Query APIs with the `_outer` suffix will include non-matching keys in the output. See
 * the relevant API documentation for more information. 
 *
 * Typical associative container query APIs like `retrieve` look up values by solely by key, e.g., `count` for a Key `k` will 
 * count all values whose associated key `k'` matches `k` as determined by `key_equal(k, k')`. In some cases, one 
 *  may want to consider both key _and_ value when determining if a key-value pair should contribute to the output.
 * `static_multimap` supports this use case with APIs prefixed with `pair_`, e.g., `pair_count` is given a key-value pair 
 * `{k,v}` and only counts key-value pairs, `{k', v'}`, in the map where `pair_equal({k,v}, {k', v'})` is true. See the relevant
 * API documentation for more information. 
```